src/mem/pool.h

```diff
@@ -32,14 +36,88 @@ namespace snmalloc
     constexpr PoolState() = default;
   };
 
-  template<typename T>
+  /**
+   * Helper class used to instantiate a global PoolState.
+   *
+   * SingletonPoolState::pool is the default provider for the PoolState within
+   * the Pool class.
+   */
+  template<typename T, typename SharedStateHandle>
+  class SingletonPoolState
+  {
+    static inline PoolState<T> state;
+    /**
+     * Thread-local initialization marker for a contention-free way to skip
+     * initialization. SharedStateHandle::ensure_init allows for multiple calls,
+     * but it uses a reentrency-safe check which is more expensive.
+     */
+    static thread_local inline bool initialized;
```

Comment:
Could you add an ecplicit `{false}` here?
```suggestion
    static thread_local inline bool initialized{false};
```

Reply:
I like the contention free check here.  We should perhaps use this trick elsewhere.

Reply:
An alternative suggested to me was to use SharedStateHandle::is_initialized, which looks at the global flag, but does not acquire a lock to do so. Would you prefer that one instead?

Reply:
I am happy with what you have.